variables.tf

```
@@ -0,0 +1,191 @@
+variable "public_ip_config" {
+  type = object({
+      enabled           = bool
+      allocation_method = optional(string, "Static")
+  })
+  default = {
+    enabled = false
+  }
+  validation {
+    condition     =  contains(["Static","Dynamic"], var.public_ip_config.allocation_method)
+    error_message = "Allocation method must be Static or Dynamic"
+  }
+  description = <<-DOC
+  ```
+    enabled: Optionally select true if a public ip should be created. Defaults to false.
+    allocation_method: The allocation method of the public ip that will be created. Defaults to static.      
+  ```
+  DOC
+} 
+
+# nsg needs to be an object to use the count object in main.tf. 
+variable "nic_config" {
+  type = object({
+      private_ip  = optional(string)
+      dns_servers = optional(list(string))
+      nsg_name    = optional(string, "")
+      nsg_rg_name = optional(string, "")
+  })
+  default = {}
+  description = <<-DOC
+  ```
+    private_ip: Optioanlly specify a private ip to use. Otherwise it will  be allocated dynamically.
+    dns_servers: Optionally specify a list of dns servers for the nic.
+    nsg_name: Optinally specify the name of a network security group that will be assigned to the nic.
+    nsg_rg_name: Optinally specify the RG name of a network security group that will be assigned to the nic.
+    nsg_id: Optinally specify the id of a network security group that will be assigned to the nic.
+  ```
+  DOC
+}
+
+variable "subnet" {
+  type               = object ({
+    id               = string
+    address_prefixes = list(string)
+  })
+  description = "The variable takes the subnet as input and takes the id and the address prefix for further configuration."
+}
+
+variable "virtual_machine_config" {
+  type = object({
+      hostname                  = string
+      size                      = string
+      location                  = string
+      zone                      = optional(string, "")
+      admin_username            = optional(string, "loc_sysadmin")
+      os_sku                    = optional(string, "gen2")
+      os_offer                  = optional(string, "sles-15-sp4")
+      os_version                = optional(string, "2023.02.05")
+      os_publisher              = optional(string, "SUSE")
+      os_disk_name              = optional(string, "OsDisk_01")
+      os_disk_caching           = optional(string, "ReadWrite")
+      os_disk_size_gb           = optional(number, 64)
```

```suggestion
      os_disk_size_gb           = optional(number)
```
If we don't specify a default value then the image will select one automatically for us that is always big enough

---

Recommendation applied.